### ISSUE TITLE
Elements CLI Preview

### ISSUE THREAD

**Issue by marbemac** [stoplightio/elements#765]
# Intro & Goal

As a non-Studio user, working in Vim or any other text editor, I want to see how my docs will look in Elements without having to make an index.html file and embed the component somewhere. Support in my text editor would be amazing, but as a fall-back having a CLI tool to use would be ok too.

## Who's it for?

- Any OpenAPI user who wants to "preview" their API refererence docs without an account, or having to push or upload things.

- Swagger UI / ReDoc users who want something nicer and more functionality to look at whilst building OpenAPI in their favourite not-Stoplight editor.

- Elements users trialing elements out.

## Why build it?

- A simple CLI exists for every popular API documentation tool: ReDoc, Swagger UI, Spectacle

- Let people play around with elements without needing to make somewhere to embed the component

## What is it?

- Runs a server instance which is just the API component embedded in a `index.html` file and served up on `localhost:8XXX/` so that folks can open it in their browser.

- `elements preview openapi.yaml`to autodetect type of thing to preview. For now, only APIs are supported. In future we can offer`elements preview --api=openapi.yaml`to be more specific, but dont need it yet.

- `--watch` option to have it reload when changes are made to the files its looking at (wont work for project, its ok just ignore it)

- `elements export openapi.yaml -o foo.html` option to leverage the existing Static Export we created elsewhere without forcing users to download swagger-cli or write JavaScript.he server.

- Embed basic *anonymous* tracking in place to get a sense of usage. Dealers choice on exactly what but something we generally use elsewhere too.

## Out of Scope

We will not be offering previews for projects, articles, or models at this time, only APIs.

## Key Success Criteria

- Double the Stoplight CLI downloads (from 8,000 to 16,000) within 2 months of release

<p></p>

<p></p>

**Comment by steve-nay-sage** [comment]
As a technical writer, I was really excited to start reading about this feature. Then I got to the Out of Scope section: 

"We will not be offering previews for projects, articles, or models at this time, only APIs." 

That makes this just another OpenAPI definition formatter. Why bother? There are already several of those available from other developers; it's not as if people are blocked by not being able to find a tool that lets them preview a pretty version of their OpenAPI definition on their local machine. 

If Elements is for presenting API documentation, and your third identified users for this feature are "Elements users trialing elements out," then only formatting OpenAPI files leaves a lot out, doesn't it? 

Right now (8/10/21) there is no way to "try" Elements except on a public project, but why would you want to experiment with something in public? That doesn't make much sense. If you could build this tool to give most of Elements functionality on a local machine, it would be really helpful. And by "most" I mean including articles, models, and custom TOC.

**Comment by colinmollenhour** [comment]
This would be really handy, having to mix tools from Redocly and Swagger to get this functionality is a sub-par experience. My use-cases:

- Setup review apps in the project CI environment
- Support hot-reload for dev - allows me to use OAS features that Stoplight Studio doesn't support such as $ref anywhere - a pattern learned from Redocly

**Comment by schroedan** [comment]
Hello everyone, we from the [skriptfabrik](https://github.com/skriptfabrik) didn't want to wait any longer and have crafted our own [Elements CLI](https://github.com/skriptfabrik/elements-cli).
We published it today and hope that it is in the interest of Stoplight and the community and that it will be accepted by all.
The documentation is still a bit poor. We would be happy if we can count on your support.
Here are the key features in a nutshell:

* export as stand-alone HTML version with configurable web component
* preview via HTTP server + CORS proxy + livereload/watcher
* Elements CLI as [NPM package](https://www.npmjs.com/package/@skriptfabrik/elements-cli)
* Elements CLI as [Docker image](https://hub.docker.com/r/skriptfabrik/elements-cli)

Please check it out and feel free to give us feedback or create issues on GitHub.

Cheers!

**Comment by philsturgeon** [comment]
@schroedan brilliant work, thanks for contributing this. It’s what open-source is all about! 🙌🏻
